### ISSUE TITLE
The virtual machine doesn't work properly under libvirt

### ISSUE THREAD

**Issue by marmistrz** [maemo-leste/bugtracker#291]
Steps to reproduce:
1. Download the qcow2 image of the Leste VM and put it into `/var/lib/libvirt/images`.
2. Open virt-manager, import the virtual machine. I gave it 4096M RAM and 2 cores. Select _Debian 9_ as the preset for the operating system.

The following issues happen:
1. The boot fails with _gave up waiting for the root file system device_. It appears that the VM doesn't support disk access via virtio - the system boots after I change the disk bus from `virtio` to `ide` or `scsi/sata`.
2. Even if I boot the VM using the previous workaround, `hildon-desktop` doesn't really show anything - it either shows the menu button in the top left corner or just a black screen. Changing video from QXL to VGA doesn't change anything.

If I use the _generic_ preset, then only issue 2 occurs.

**Comment by MerlijnWajer** [comment]
Well, the Debian version we use may not support virtio. Virtio for drives or video will not yet work, I think. Are you sure you're not using virtio for video output? Try VNC?

**Comment by marmistrz** [comment]
In virt-manager there's VGA, QXL and virtio available, neither of them works, even though the guest X.org driver for QXL seems to be installed.

**Comment by MerlijnWajer** [comment]
What's your resolution?

**Comment by MerlijnWajer** [comment]
I think that maybe I haven't submitted the fixed hildon-desktop to `leste` yet, only `leste-devel`.

**Comment by MerlijnWajer** [comment]
Also, the modesetting driver also has problems. You can probably get something on the screen if you type `chvt 1; chvt 2`. Put the screen still won't update. I think /maybe/ in the droid4 component we have a fix for this, but I am not sure if's the same modesetting problem.

**Comment by marmistrz** [comment]
xrandr shows 1024x768, max possible is 1920x1080. And I don't actually see why devuan wouldn't support virtio - this should be in the kernel. But I can see that many of virtio drivers are build as `m`, not `y` and aren't available in initramfs, so this is most likely the problem with disk devices.

**Comment by dderby** [comment]
The VM works fine under libvirt.  It _does_ support virtio, but there is a problem in that the cmdline in GRUB has the root device hard-coded to `/dev/sda1`, but virtio uses `/dev/vda1` instead.

As a workaround, you can interrupt GRUB and press `e`, then edit the cmdline to point the root device to `/dev/vda1`.  Once booted succefully, you can run `update-grub` to fix grub.cfg which will replace /dev/sda1 with the UUID. 

To fix this properly, we should set up GRUB to point root device to the UUID or PARTUUID when creating the image.

Any thoughts @parazyd?

**Comment by goetzk** [comment]
I booted the most recent (December) VM under virtio and it passed grub - it failed during booting ( I raised this via email but didn't provide any information there either https://lists.dyne.org/lurker/message/20201230.221458.9b3a48b7.en.html).
This appears to be due to using uuid for boot "out of the box".

I had an issue with the keyboard but don't think I had any display issues (at least, not that I noticed).

**Comment by parazyd** [comment]
I will try to set up the image to use UUID in fstab.

**Comment by parazyd** [comment]
@goetzk @marmistrz @dderby 
I implemented this patch: https://github.com/parazyd/libdevuansdk/commit/e2876ea69f31cfd150cf7577e30fb1953de53159
Also uploaded a new image which you can test: https://maedevu.maemo.org/images/virtual-machines/20210113/

I managed to boot this just fine when I made qemu use virtio with `-drive file=leste.qcow2,if=virtio` and the partitions showed up as `/dev/vda`.

Please let me know when you test this image :)

**Comment by dderby** [comment]
I tested this with Virtual Machine Manager (virt-manager).  I added a new VM pointing to the new image and keeping all default settings.  The VM now boots out-of-the-box.  GRUB is now using the UUID.  `/etc/fstab` also lists the UUID.

I'm not sure why @goetzk needed to do this: https://leste.maemo.org/index.php?title=Virtual_Machine&curid=60&diff=910&oldid=761
This was never needed for me, even before your patch.  Maybe we should let him test it before closing this out.
